main.py

```diff
@@ -384,44 +386,33 @@ def check_alien_flows(self, switch):
                         f"Flow: {flow}"
                     )
 
-    # pylint: disable=attribute-defined-outside-init
-    def _delete_matched_flows(self, flow_dict, _flow_id, _match_id, switch):
-        """Try to delete matching stored flows given a flow dict."""
-        cookie = flow_dict.get("cookie", 0)
-        stored_flows = self.flow_controller.get_flows_by_cookie(switch.id, cookie)
-        version = switch.connection.protocol.version
-        flow_ids = []
-        for flow in stored_flows:
-            if match_flow(flow_dict, version, flow["flow"]):
-                flow_ids.append(flow["flow_id"])
-        if flow_ids:
-            self.flow_controller.delete_flows_by_ids(flow_ids)
-
-    def _upsert_flow(self, flow_dict, flow_id, match_id, switch):
-        """Try to add a flow dict in the store idempotently."""
-        flow_dict = {**{"flow": flow_dict}, **{"switch": switch.id, "flow_id": flow_id}}
-        self.flow_controller.upsert_flow(match_id, flow_dict)
-
-    def _store_changed_flows(self, command, flow_dict, flow_id, match_id, switch):
-        """Store changed flows.
+    def delete_matched_flows(self, flow_dicts, switches: dict) -> None:
+        """Try to delete many matched stored flows given flow_dicts for switches.
 
-        Args:
-            command: Flow command to be installed
-            flow: flow dict to be stored
-            flow_id: corresponding FlowMod id (used for indexing)
-            match_id: corresponding FlowMod match_id (used for indexing)
-            switch: Switch target
+        This deletion tries to minimize DB round trips, it aggregates all
+        included cookies grouped by dpids, and then at the runtime it performs
+        a non strict match iterating over the flows if they haven't been deleted yet.
+        If flows are matched, they will be bulk deleted.
         """
-        cmd_handlers = {
-            "add": self._upsert_flow,
-            "delete": self._delete_matched_flows,
-        }
-        if command not in cmd_handlers:
-            raise ValueError(
-                f"Invalid command: {command}, supported: {list(cmd_handlers.keys())}"
-            )
-        return cmd_handlers[command](flow_dict, flow_id, match_id, switch)
+        flow_ids_to_delete = set()
+        cookies = list({int(value.get("cookie", 0)) for value in flow_dicts})
+        for dpid, stored_flows in self.flow_controller.get_flows_by_cookies(
+            list(switches.keys()), cookies
+        ).items():
+            for flow_dict in flow_dicts:
+                for stored_flow in stored_flows:
+                    if stored_flow["flow_id"] in flow_ids_to_delete:
+                        continue
+                    if match_flow(
```

Comment:
In the future, we could also offload this non strict match to the DB, the performance is great now, so probably not worth mapping this enhancement or trying to do this now. Having a full compliant OpenFlow non strict match offloaded would be challenging too (especially with masks), but depending in the future if we reduce the scope and support the major types of matches on certain fields that should be doable with indexes and some query expressions that MongoDB provides.